h2o-automl/src/main/java/ai/h2o/automl/targetencoding/BroadcastJoinForTargetEncoder.java

```diff
@@ -0,0 +1,168 @@
+      return _encodingDataMapPerNode;
+    }
+
+    public IcedHashMap<CompositeLookupKey, EncodingData> getEncodingDataMapFinal() {
+      return _encodingDataMapFinal;
+    }
+  }
+
+  /**
+   * 
+   * @param leftFrame frame that we want to keep order of
+   * @param leftCatColumnsIdxs indices of the categorical columns from `leftFrame` we want to use to calculate encodings
+   * @param leftFoldColumnIdx index of the fold column from `leftFrame` or `-1` if we don't use folds
+   * @param broadcastedFrame supposedly small frame that we will broadcast to all nodes and use it as lookup table for joining
+   * @param rightCatColumnsIdxs indices of the categorical columns from `broadcastedFrame` we want to use to calculate encodings 
+   * @param rightFoldColumnIdx index of the fold column from `broadcastedFrame` or `-1` if we don't use folds
+   * @return
+   */
+  static Frame join(Frame leftFrame, int[] leftCatColumnsIdxs, int leftFoldColumnIdx, Frame broadcastedFrame, int[] rightCatColumnsIdxs, int rightFoldColumnIdx) {
+    int numeratorIdx = broadcastedFrame.find(TargetEncoder.NUMERATOR_COL_NAME);
+    int denominatorIdx = broadcastedFrame.find(TargetEncoder.DENOMINATOR_COL_NAME);
+
+    IcedHashMap<CompositeLookupKey, EncodingData> encodingDataMap = new FrameWithEncodingDataToHashMap(rightCatColumnsIdxs[0], rightFoldColumnIdx, numeratorIdx, denominatorIdx)
+            .doAll(broadcastedFrame)
+            .getEncodingDataMap();
+    new BroadcastJoiner(leftCatColumnsIdxs, leftFoldColumnIdx, encodingDataMap).doAll(leftFrame);
+    return leftFrame;
+  }
+
+  static class BroadcastJoiner extends MRTask<BroadcastJoiner> {
+    int _categoricalColumnIdx, _foldColumnIdx;
+    IcedHashMap<CompositeLookupKey, EncodingData> _encodingDataMap;
+
+    BroadcastJoiner(int[] categoricalColumnsIdxs, int foldColumnIdx, IcedHashMap<CompositeLookupKey, EncodingData> encodingDataMap) {
+      assert categoricalColumnsIdxs.length == 1 : "Only single column target encoding is supported for now";
+
+      this._categoricalColumnIdx = categoricalColumnsIdxs[0];
+      this._foldColumnIdx = foldColumnIdx;
+      this._encodingDataMap = encodingDataMap;
+    }
+
+    @Override
+    public void map(Chunk[] cs) {
+      Chunk categoricalChunk = cs[_categoricalColumnIdx];
+      int numOfVecs = cs.length;
+      Chunk num = cs[numOfVecs - 2];
+      Chunk den = cs[numOfVecs - 1];
+      for (int i = 0; i < num.len(); i++) {
+        long levelValue = categoricalChunk.at8(i);
+        String factor = categoricalChunk.vec().factor(levelValue);
+        
+        int foldValue = -1;
+        if(_foldColumnIdx != -1) {
+          long foldValueFromVec = cs[_foldColumnIdx].at8(i);
+          assert foldValueFromVec <= Integer.MAX_VALUE && foldValueFromVec >= 0 : FOLD_VALUE_IS_OUT_OF_RANGE_MSG;
+          foldValue = (int) foldValueFromVec;
+        }
+        
+        CompositeLookupKey lookupKey = new CompositeLookupKey(factor, foldValue);
```

Comment:
We usually avoid allocating new objects on a per-row level if it is not necessary. This could be done by mutating CompositeLookupKey. Not necessary just a remark ;)

Reply:
@michalkurka  echoes of my Scala background -> generating tons of objects is not a problem as long as everything is immutable. But you are right, I will rewrite it into a mutable version.